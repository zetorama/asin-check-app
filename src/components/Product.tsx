import React, { Component } from 'react'
import { ListGroup, ListGroupItem, Button, Alert } from 'reactstrap'

import { GrabbedProduct, GrabStatus } from '../product'

const InfoItem = ({ label, children }: { label: string; children: React.ReactNode }) => (
	<ListGroupItem>
		<strong className='mr-2'>{label}:</strong>
		{children}
	</ListGroupItem>
)

export interface ProductProps {
	product: GrabbedProduct
}

export class Product extends Component<ProductProps> {
	render() {
		const { product } = this.props
		if (!product) return null

		return (
			<div className='well'>
				<h1>
					ASIN:
					<code>{product.asin}</code>
				</h1>
				{product.data && product.data.title && <h2>{product.data.title}</h2>}
				<ListGroup>
					{product.data ? this.renderDetails() : this.renderExcuse()}
					<InfoItem label='Updated On'>
						{String(product.updatedOn)}{' '}
						<Button color='primary' size='sm' outline>
							Refresh
						</Button>
					</InfoItem>
				</ListGroup>
			</div>
		)
	}

	renderExcuse() {
		const { status } = this.props.product
		switch (status) {
			case GrabStatus.Timeout:
				return (
					<ListGroupItem>
						<Alert color='warning'>Sync was canceled due to timeout. Please, try again</Alert>
					</ListGroupItem>
				)
			case GrabStatus.Nomatch:
				return (
					<ListGroupItem>
						<Alert color='warning'>Product with such ASIN wasn't found on Amazon. Sorry for that</Alert>
					</ListGroupItem>
				)
			case GrabStatus.Fail:
				return (
					<ListGroupItem>
						<Alert color='danger'>There was some fatal error during last sync. I wish I had logs</Alert>
					</ListGroupItem>
				)
		}
	}

	renderDetails() {
		const { data } = this.props.product
		if (!data) return null

		return (
			<>
				<InfoItem label='Category'>{data.categories.join(' › ')}</InfoItem>
				<InfoItem label='Best Sellers Rank'>{this.renderRank()}</InfoItem>
				<InfoItem label='Product Dimensions'>{this.renderDimensions()}</InfoItem>
			</>
		)
	}

	renderDimensions() {
		const { data } = this.props.product
		if (!data) return null

		return data.dimensions || '—'
	}

	renderRank() {
		const { data } = this.props.product
		if (!data) return null

		if (!data.rank) {
			return '—'
		}

		return data.rank.map((rank) => (
			<>
				<br />
				{rank}
			</>
		))
	}
}

export default Product
