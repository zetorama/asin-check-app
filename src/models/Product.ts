import { getParent, types, Instance, SnapshotIn } from 'mobx-state-tree'

import { StoreModel } from './Store'

export enum ProductStatus {
    Init = 'init',
    Queue = 'queue',
    Fetch = 'fetch',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export const ProductDetails = types.model('ProductDetails', {
    title: types.string,
    categories: types.array(types.string),
    dimensions: types.maybe(types.string),
    rank: types.maybe(types.array(types.string)),
    rating: types.maybe(types.string),
})

export type ProductAsin = string

export type ProductModel = Instance<typeof Product>
export type ProductSnapshot = SnapshotIn<typeof Product>

export const Product = types
    .model('Product', {
        asin: types.identifier,
        status: types.enumeration<ProductStatus>('ProductStatus', Object.values(ProductStatus)),
        data: types.maybe(types.frozen(ProductDetails)),

        createdOn: types.maybe(types.string),
        updatedOn: types.maybe(types.string),
        isViewed: false,
    })
    .views((self) => ({
        get isLoading(): boolean {
            return (
                self.status === ProductStatus.Init ||
                self.status === ProductStatus.Queue ||
                self.status === ProductStatus.Fetch
            )
        },
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    }))
    .actions((self) => ({
        toggleViewed(setAsViewed: boolean = !self.isViewed) {
            self.isViewed = setAsViewed
        },
    }))

export default Product
