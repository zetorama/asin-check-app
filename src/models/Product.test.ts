import { getSnapshot } from 'mobx-state-tree'
import { Product, ProductStatus } from './Product'

const dummyProduct = (status: ProductStatus = ProductStatus.Success) => ({
    asin: 'XXX',
    status,
    data: {
        title: 'xxx aaa',
        categories: ['xxx', 'aaa'],
        dimensions: '1 x 2 x 3',
        rank: ['1', '2', '3'],
        rating: 'ok',
    },
})

describe('Product model', () => {
    it('can create a model', () => {
        const model = Product.create(dummyProduct())

        expect(getSnapshot(model)).toMatchSnapshot()
    })

    it('switches status', () => {
        const model = Product.create(dummyProduct())
        model.setStatus(ProductStatus.Fail)
        expect(model.status).toBe(ProductStatus.Fail)
    })

    it('toggles viewed flag', () => {
        const model = Product.create(dummyProduct())
        expect(model.isViewed).toBe(false)
        model.toggleViewed()
        expect(model.isViewed).toBe(true)
        model.toggleViewed()
        expect(model.isViewed).toBe(false)
        model.toggleViewed(false)
        expect(model.isViewed).toBe(false)
        model.toggleViewed(true)
        expect(model.isViewed).toBe(true)
    })

    it('calculates loading flag', () => {
        const model = Product.create(dummyProduct(ProductStatus.New))
        expect(model.isLoading).toBe(true)
        model.setStatus(ProductStatus.InQueue)
        expect(model.isLoading).toBe(true)
        model.setStatus(ProductStatus.Refreshing)
        expect(model.isLoading).toBe(true)
        model.setStatus(ProductStatus.Success)
        expect(model.isLoading).toBe(false)
        model.setStatus(ProductStatus.Timeout)
        expect(model.isLoading).toBe(false)
        model.setStatus(ProductStatus.Nomatch)
        expect(model.isLoading).toBe(false)
        model.setStatus(ProductStatus.Fail)
        expect(model.isLoading).toBe(false)
        model.setStatus(ProductStatus.New)
        expect(model.isLoading).toBe(true)
    })
})
