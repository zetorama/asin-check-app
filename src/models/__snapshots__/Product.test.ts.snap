// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product model can create a model 1`] = `
Object {
  "asin": "XXX",
  "createdOn": undefined,
  "data": Object {
    "categories": Array [
      "xxx",
      "aaa",
    ],
    "dimensions": "1 x 2 x 3",
    "rank": Array [
      "1",
      "2",
      "3",
    ],
    "rating": "ok",
    "title": "xxx aaa",
  },
  "isViewed": false,
  "status": "success",
  "updatedOn": undefined,
}
`;
